perception/autoware_probabilistic_occupancy_grid_map/lib/utils/utils_kernel.cu

```diff
@@ -0,0 +1,296 @@
+  unsigned int y0{};
+  const float ox{source_x};
+  const float oy{source_y};
+
+  if (!worldToMap(ox, oy, x0, y0, origin_x, origin_y, resolution_inv, size_x, size_y)) {
+    return;
+  }
+
+  // we can pre-compute the endpoints of the map outside of the inner loop... we'll need these later
+  const float resolution = 1.0 / resolution_inv;
+  const float map_end_x = origin_x + size_x * resolution;
+  const float map_end_y = origin_y + size_y * resolution;
+
+  float wx = target_x;
+  float wy = target_y;
+
+  // now we also need to make sure that the endpoint we're ray-tracing
+  // to isn't off the costmap and scale if necessary
+  const float a = wx - ox;
+  const float b = wy - oy;
+
+  // the minimum value to raytrace from is the origin
+  if (wx < origin_x) {
+    const float t = (origin_x - ox) / a;
+    wx = origin_x;
+    wy = oy + b * t;
+  }
+  if (wy < origin_y) {
+    const float t = (origin_y - oy) / b;
+    wx = ox + a * t;
+    wy = origin_y;
+  }
+
+  // the maximum value to raytrace to is the end of the map
+  if (wx > map_end_x) {
+    const float t = (map_end_x - ox) / a;
+    wx = map_end_x - .001;
+    wy = oy + b * t;
+  }
+  if (wy > map_end_y) {
+    const float t = (map_end_y - oy) / b;
+    wx = ox + a * t;
+    wy = map_end_y - .001;
+  }
+
+  // now that the vector is scaled correctly... we'll get the map coordinates of its endpoint
+  unsigned int x1{};
+  unsigned int y1{};
+
+  // check for legality just in case
+  if (!worldToMap(wx, wy, x1, y1, origin_x, origin_y, resolution_inv, size_x, size_y)) {
+    return;
+  }
+
+  constexpr unsigned int cell_raytrace_range = 10000;  // large number to ignore range threshold
+  raytraceLine(x0, y0, x1, y1, cell_raytrace_range, 0, size_x, cost, costmap_tensor);
+}
+
+__global__ void transformPointCloudKernel(
+  std::uint8_t * points, std::size_t num_points, std::size_t points_step,
+  const Eigen::Matrix3f & rotation, const Eigen::Vector3f & translation)
+{
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= num_points) {
+    return;
+  }
+
+  Eigen::Map<Eigen::Vector3f> point_map(reinterpret_cast<float *>(points + idx * points_step));
+  point_map = rotation * point_map + translation;
+}
+
+__global__ void copyMapRegionKernel(
+  const std::uint8_t * source_map, unsigned int sm_lower_left_x, unsigned int sm_lower_left_y,
+  unsigned int sm_size_x, std::uint8_t * dest_map, unsigned int dm_lower_left_x,
+  unsigned int dm_lower_left_y, unsigned int dm_size_x, unsigned int region_size_x,
+  unsigned int region_size_y)
+{
+  const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= region_size_x * region_size_y) return;
+
+  const int region_y = idx / region_size_x;
+  const int region_x = idx % region_size_x;
+
+  const int sm_index =
+    sm_lower_left_y * sm_size_x + sm_lower_left_x + region_y * sm_size_x + region_x;
+  const int dm_index =
+    dm_lower_left_y * dm_size_x + dm_lower_left_x + region_y * dm_size_x + region_x;
+
```

Comment:
[imho]
for safety

```
if(dm_index >= dm_size_x * dm_size_y) return;
```

Reply:
Addressed in [ea46130](https://github.com/autowarefoundation/autoware.universe/pull/9542/commits/ea461308f5c992e769218cfa56da90e90b36f147)